File: backend_nodejs/src/presentation/handler/getTodoHandler.ts
import type { NextFunction, Request, Response } from "express";
import z from "zod";
import type { GetTodoUsecase } from "@/application_service/usecase/getTodoUsecase";
import { RepositoryNotFoundError } from "@/domain/repository/errors";
import {
  BadRequestError,
  NotFoundError,
} from "@/presentation/middleware/errorHandler";

const getTodoRequestSchema = z.object({
  params: z.object({
    id: z.uuid({ version: "v7" }),
  }),
});

type GetTodoResponse = {
  id: string;
  title: string;
  description: string | null;
  completed: boolean;
};

export class GetTodoHandler {
  constructor(private getTodoUsecase: GetTodoUsecase) {}

  handle = async (
    req: Request,
    res: Response<GetTodoResponse>,
    next: NextFunction,
  ) => {
    try {
      const parsedReq = getTodoRequestSchema.parse(req);
      const result = await this.getTodoUsecase.execute({
        id: parsedReq.params.id,
      });
      res.json({
        id: result.todo.id,
        title: result.todo.title,
        description: result.todo.description,
        completed: result.todo.isCompleted,
      });
    } catch (error) {
      switch (true) {
        case error instanceof z.ZodError: {
          const message = error.issues
            .map((issue) => `${issue.path}: ${issue.message}`)
            .join(", ");
          return next(
            new BadRequestError(`Invalid request parameters: [${message}]`),
          );
        }
        case error instanceof RepositoryNotFoundError:
          return next(new NotFoundError("Todo not found"));
      }
      return next(error);
    }
  };
}

File: backend_python/src/todo_api/presentation/handler/get_todo_handler.py
from pydantic import BaseModel

from todo_api.application_service.usecase.get_todo_usecase import (
    GetTodoUsecase,
    GetTodoUsecaseInput,
)
from todo_api.domain.repository.errors import RepositoryNotFoundError
from todo_api.presentation.middleware.error_handler import NotFoundError
from todo_api.utils.uuid import UUID7


class GetTodoResponse(BaseModel):
    id: str
    title: str
    description: str | None = None
    completed: bool


class GetTodoHandler:
    def __init__(self, get_todo_usecase: GetTodoUsecase) -> None:
        self.get_todo_usecase = get_todo_usecase

    def handle(self, id: UUID7) -> GetTodoResponse:
        try:
            result = self.get_todo_usecase.execute(GetTodoUsecaseInput(id=id))
        except RepositoryNotFoundError:
            raise NotFoundError(message="Todo not found")
        return GetTodoResponse(
            id=str(result.todo.id),
            title=result.todo.title,
            description=result.todo.description,
            completed=result.todo.completed,
        )
